### PR TITLE
MGMT-8100: e2e test for late unbind

### PR DESCRIPTION
**Pull request by slaviered**

### CODE REVIEW

**Comment by rollandf** [comment]
Logic looks good. LGTM
@eliorerz can you please take a look also.

**Comment by eliorerz** [comment]
/hold 
@slaviered Add CI job (optional) for this test so we will be able to run and test in in your PR

File: discovery-infra/tests/test_kube_api.py [hunk 1]
```diff
@@ -164,14 +175,19 @@ def _set_agent_cluster_install_machine_cidr(agent_cluster_install, nodes):
         agent_cluster_install.set_machinenetwork(machine_cidr)
 
     @classmethod
```
Comment (eliorerz) [inline discovery-infra/tests/test_kube_api.py:177]:
```suggestion
    @classmethod
    @JunitTestCase()
```

File: discovery-infra/tests/test_kube_api.py [hunk 2]
```diff
@@ -164,14 +175,19 @@ def _set_agent_cluster_install_machine_cidr(agent_cluster_install, nodes):
...
-    def _late_binding_install(cls, cluster_deployment, agent_cluster_install, agents, nodes, is_ipv4):
+    def _late_binding_install(
+        cls, cluster_deployment, agent_cluster_install, agents, nodes, is_ipv4, hold_install=False
```
Comment (eliorerz) [inline discovery-infra/tests/test_kube_api.py:179]: Fix missing type annotations

File: discovery-infra/tests/test_kube_api.py [hunk 3]
```diff
@@ -204,7 +236,13 @@ def test_kube_api_late_binding_ipv4_highly_available(
 
 @JunitTestCase()
 def kube_api_test_prepare_late_binding_cluster(
```
Comment (eliorerz) [inline discovery-infra/tests/test_kube_api.py:238]: Same here. Add type hint annotations

**Review by eliorerz — CHANGES_REQUESTED** [review]

File: discovery-infra/test_infra/helper_classes/kube_helpers/agent.py [hunk 4]
```diff
@@ -137,6 +137,16 @@ def bind(self, cluster_deployment) -> None:
         )
         logger.info(f"Bound agent {self.ref} to cluster_deployment {cluster_deployment.ref}")
 
+    @classmethod
+    def wait_for_agents_to_be_bound(
```
Comment (filanov) [inline discovery-infra/test_infra/helper_classes/kube_helpers/agent.py:141]: Is that used?

File: discovery-infra/test_infra/helper_classes/kube_helpers/agent.py [hunk 5]
```diff
@@ -187,6 +209,7 @@ def wait_till_all_agents_are_in_status(
         agents: List["Agent"],
         statusType: str,
         timeout,
+        status="True",
```
Comment (filanov) [inline discovery-infra/test_infra/helper_classes/kube_helpers/agent.py:212]: Maybe should rename to statusVal or something like that?

File: discovery-infra/tests/test_kube_api.py [hunk 6]
```diff
@@ -184,6 +207,22 @@ def test_kube_api_late_binding_ipv4_single_node(self, unbound_single_node_cluste
...
+        self, unbound_single_node_cluster_hold_installation, unbound_single_node_infraenv
+    ):
+        infra_env, nodes = unbound_single_node_infraenv
+        cluster_deployment, agent_cluster_install, cluster_config = unbound_single_node_cluster_hold_installation
+
+        agents = infra_env.wait_for_agents(len(nodes))
+        self._late_binding_install(
+            cluster_deployment, agent_cluster_install, agents, nodes, is_ipv4=True, hold_install=True
+        )
+
```
Comment (filanov) [inline discovery-infra/tests/test_kube_api.py:222]: Maybe need to check that host is bound? `wait_for_agents_to_be_bound`

File: discovery-infra/test_infra/helper_classes/kube_helpers/agent.py (hunk 4 above)
Comment (slaviered) [reply discovery-infra/test_infra/helper_classes/kube_helpers/agent.py:141]: see the fix

File: discovery-infra/tests/test_kube_api.py (hunk 6 above)
Comment (slaviered) [reply discovery-infra/tests/test_kube_api.py:223]: fixed

File: src/assisted_test_infra/test_infra/helper_classes/kube_helpers/agent.py [hunk 7]
```diff
@@ -187,6 +209,7 @@ def wait_till_all_agents_are_in_status(
         agents: List["Agent"],
         statusType: str,
```
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/helper_classes/kube_helpers/agent.py:210]: Argument name should be lowercase

File: src/assisted_test_infra/test_infra/helper_classes/kube_helpers/agent.py [hunk 8]
```diff
@@ -195,7 +218,7 @@ def wait_till_all_agents_are_in_status(
             lambda: Agent.are_agents_in_status(
                 agents,
                 statusType,
```
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/helper_classes/kube_helpers/agent.py:220]: Argument name should be lowercase

File: src/tests/test_kube_api.py [hunk 9]
```diff
@@ -203,8 +243,14 @@ def test_kube_api_late_binding_ipv4_highly_available(
...
 def kube_api_test_prepare_late_binding_cluster(
-    kube_api_context, cluster_config: ClusterConfig, num_controlplane_agents, *, proxy_server=None, is_ipv4=True
-):
+    kube_api_context: KubeAPIContext,
+    cluster_config: ClusterConfig,
+    num_controlplane_agents: int,
+    *,
+    proxy_server=None,
+    is_ipv4=True,
+    hold_install=False,
```
Comment (eliorerz) [inline src/tests/test_kube_api.py:252]:
```suggestion
    hold_installation=False,
```

**Review by eliorerz — CHANGES_REQUESTED** [review]

File: src/tests/test_kube_api.py (hunk 9 above)
Comment (slaviered) [reply src/tests/test_kube_api.py]: fixed

File: src/assisted_test_infra/test_infra/helper_classes/kube_helpers/agent.py (hunk 7 above)
Comment (slaviered) [reply src/assisted_test_infra/test_infra/helper_classes/kube_helpers/agent.py:210]: A major change in the existing code (all relevant and existing functions use statusType). Out of scope for this PR

**Review by eliorerz — APPROVED** [review]
LGTM

**Pull request merged**